python/ray/data/block.py

```diff
@@ -252,6 +252,10 @@ def iter_rows(self) -> Iterator[T]:
         """Iterate over the rows of this block."""
         raise NotImplementedError
 
+    def get_keys(self, key: KeyFn) -> Optional[np.ndarray]:
```

Comment:
it seems a little weird to me to have a `get_keys()` returning just `np.ndarray` defined in `BlockAccessor` interface.

Is it possible to add more logic into existing `to_numpy` function?

```
def to_numpy(
    self, columns: Optional[Union[str, List[str]]] = None
) -> Union[np.ndarray, Dict[str, np.ndarray]]:
```

to

```
def to_numpy(
    self, columns: Optional[Union[KeyFn, List[str]]] = None
) -> Union[np.ndarray, Dict[str, np.ndarray]]:
```

Reply:
For the added interface:

```
def get_keys(self, key: KeyFn) -> Optional[np.ndarray]:
```

For a given `block`, the function gets the required key value based on `KeyFun`, in numpy ndarray representation.
What if in the future, we need get the required key value in another representation e.g. pandas dataframe?
I think numpy ndarray is just one of internal data representations. Can't we reuse the existing `to_numpy` function? It might be confusing we have both:

```
def to_numpy(self, columns: Optional[Union[str, List[str]]] = None
) -> Union[np.ndarray, Dict[str, np.ndarray]]
```

and 

```
def get_keys(self, key: KeyFn) -> Optional[np.ndarray]:
```

Reply:
get_keys() is not about in numpy conversion, that's not its role here. numpy array is just as an efficient format chosen for return value. The interesting part is the key or KeyFn, and their future evolution, not particularly the return type.

Reply:
hmm I might be miss something, for `get_keys()` aren't we trying to get `key` column from block? Essentially we are doing columns projection on table, and convert the columns to numpy format.

Would it be more flexiable if we have:

```
def select(self, keys=Union[KeyFn, List[KeyFn]]) -> Block
```

And we can do

```
block_accessor.select(self._key).to_numpy()
```

So we get benefit of evolution `keys` parameter in the future, and we are not locked in with return type. When we need to select keys in another format, we can just call `block_accessor.select(self._key).to_pandas()`.

btw [PyArrow has similar `select()` API](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.select).

Reply:
I think it's overskilled for this change. I'd actually like to not touch this at all.